compiler/rustc_hir_typeck/src/delegation.rs

```diff
@@ -0,0 +1,371 @@
+                let predicates = predicates.fold_with(&mut folder);
+                let predicates = self.tcx.arena.alloc_from_iter(predicates);
+                Some(ty::GenericPredicates { parent: None, predicates })
+            }
+            // here generics are not yet supported
+            _ => None,
+        }
+    }
+
+    fn check_constraints(&self) -> Result<(), ErrorGuaranteed> {
+        let mut ret = Ok(());
+
+        let tcx = self.tcx;
+        let span = tcx.def_span(self.def_id);
+        let sig_id = self.sig_id;
+
+        let sig_span = tcx.def_span(sig_id);
+        let mut emit = |descr| {
+            ret = Err(tcx.dcx().emit_err(crate::errors::NotSupportedDelegation {
+                span,
+                descr,
+                callee_span: sig_span,
+            }));
+        };
+
+        if let Some(node) = tcx.hir().get_if_local(sig_id)
+            && let Some(decl) = node.fn_decl()
+            && let hir::FnRetTy::Return(ty) = decl.output
+            && let hir::TyKind::InferDelegation(_, _) = ty.kind
+        {
+            emit("recursive delegation is not supported yet");
+        }
+
+        let caller_kind = fn_kind(self.tcx, self.def_id.into());
+        if caller_kind != FnKind::Free {
+            let sig_generics = self.tcx.generics_of(sig_id);
+            let parent = self.tcx.parent(self.def_id.into());
+            let parent_generics = self.tcx.generics_of(parent);
+
+            let parent_is_trait = (self.tcx.def_kind(parent) == DefKind::Trait) as usize;
+            let sig_has_self = sig_generics.has_self as usize;
+
+            if sig_generics.count() > sig_has_self || parent_generics.count() > parent_is_trait {
+                emit("early bound generics is not supported for associated delegation items");
+            }
+        }
+
+        ret
+    }
+}
+
+fn generate_error_results<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    err: ErrorGuaranteed,
+) -> ty::DelegationResults<'tcx> {
+    let sig_id = tcx.hir().delegation_sig_id(def_id);
+    let sig_len = tcx.fn_arg_names(sig_id).len();
+    let err_type = Ty::new_error(tcx, err);
+    let inputs = tcx.arena.alloc_from_iter((0..sig_len).map(|_| err_type));
+    ty::DelegationResults { inputs, output: err_type, generics: None, predicates: None }
+}
+
+pub fn resolve_delegation<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+) -> ty::DelegationResults<'tcx> {
+    let info = DelegationInfoCollector::collect(tcx, def_id);
+    let delegation_resolver = DelegationResolver::new(tcx, def_id, &info);
+    if let Err(err) = delegation_resolver.check_constraints() {
+        return generate_error_results(tcx, def_id, err);
+    }
+    let (inputs, output) = delegation_resolver.fn_sig();
+    let generics = delegation_resolver.generics_of();
+    let predicates = delegation_resolver.predicates_of();
+
+    ty::DelegationResults { inputs, output, generics, predicates }
```

Comment:
They all depend on the pass, which collects parameter definitions generic arguments. (`GenericDefsCollector`)

Reply:
Maybe that pass should be a query instead of `lower_delegation_ty`, and the later parts can be just called from `predicates_of`/etc because they don't need memoization.

Reply:
> I also suspect that the data from GenericDefsMap is already collected and available from elsewhere.

Reply:
> Maybe that pass should be a query instead of lower_delegation_ty, and the later parts can be just called from predicates_of/etc because they don't need memoization.

Maybe. I chose this option because all the logic is localized in one module and it is easy to maintain.